// @wordpress/dataviews/build-module/components/dataviews-layouts/activity/index.mjs
// packages/dataviews/src/components/dataviews-layouts/activity/index.tsx
import clsx from "clsx";
import { Spinner } from "@wordpress/components";
import { Stack } from "@wordpress/ui";
import getDataByGroup from "../utils/get-data-by-group.mjs";
import ActivityGroup from "./activity-group.mjs";
import ActivityItems from "./activity-items.mjs";
import { Fragment, jsx, jsxs } from "react/jsx-runtime";
function ViewActivity(props) {
  const { empty, data, fields, isLoading, view, className } = props;
  const hasData = data?.length;
  if (!hasData) {
    return /* @__PURE__ */ jsx(
      "div",
      {
        className: clsx({
          "dataviews-loading": isLoading,
          "dataviews-no-results": !hasData && !isLoading
        }),
        children: !hasData && (isLoading ? /* @__PURE__ */ jsx("p", { children: /* @__PURE__ */ jsx(Spinner, {}) }) : empty)
      }
    );
  }
  const wrapperClassName = clsx("dataviews-view-activity", className);
  const groupField = view.groupBy?.field ? fields.find((field) => field.id === view.groupBy?.field) : null;
  const dataByGroup = groupField ? getDataByGroup(data, groupField) : null;
  const groupedEntries = dataByGroup ? Array.from(dataByGroup.entries()) : [];
  if (hasData && groupField && dataByGroup) {
    return /* @__PURE__ */ jsx(Stack, { direction: "column", gap: "xs", className: wrapperClassName, children: groupedEntries.map(
      ([groupName, groupData]) => /* @__PURE__ */ jsx(
        ActivityGroup,
        {
          groupName,
          groupData,
          groupField,
          showLabel: view.groupBy?.showLabel !== false,
          children: /* @__PURE__ */ jsx(
            ActivityItems,
            {
              ...props,
              data: groupData
            }
          )
        },
        groupName
      )
    ) });
  }
  return /* @__PURE__ */ jsxs(Fragment, { children: [
    /* @__PURE__ */ jsx(
      "div",
      {
        className: wrapperClassName,
        role: view.infiniteScrollEnabled ? "feed" : void 0,
        children: /* @__PURE__ */ jsx(ActivityItems, { ...props })
      }
    ),
    hasData && isLoading && /* @__PURE__ */ jsx("p", { className: "dataviews-loading-more", children: /* @__PURE__ */ jsx(Spinner, {}) })
  ] });
}
export {
  ViewActivity as default
};
