// @wordpress/dataviews/build-module/components/dataform-layouts/card/index.mjs
// packages/dataviews/src/components/dataform-layouts/card/index.tsx
import {
  Button,
  Card,
  CardBody,
  CardHeader as OriginalCardHeader
} from "@wordpress/components";
import { Badge } from "@wordpress/ui";
import {
  useCallback,
  useContext,
  useEffect,
  useMemo,
  useRef,
  useState
} from "@wordpress/element";
import { chevronDown, chevronUp } from "@wordpress/icons";
import { sprintf, _n } from "@wordpress/i18n";
import { getFormFieldLayout } from "../index.mjs";
import DataFormContext from "../../dataform-context/index.mjs";
import { DataFormLayout } from "../data-form-layout.mjs";
import { DEFAULT_LAYOUT } from "../normalize-form.mjs";
import { getSummaryFields } from "../get-summary-fields.mjs";
import useReportValidity from "../../../hooks/use-report-validity.mjs";
import { jsx, jsxs } from "react/jsx-runtime";
function countInvalidFields(validity) {
  if (!validity) {
    return 0;
  }
  let count = 0;
  const validityRules = Object.keys(validity).filter(
    (key) => key !== "children"
  );
  for (const key of validityRules) {
    const rule = validity[key];
    if (rule?.type === "invalid") {
      count++;
    }
  }
  if (validity.children) {
    for (const childValidity of Object.values(validity.children)) {
      count += countInvalidFields(childValidity);
    }
  }
  return count;
}
var NonCollapsibleCardHeader = ({
  children,
  ...props
}) => /* @__PURE__ */ jsx(OriginalCardHeader, { isBorderless: true, ...props, children: /* @__PURE__ */ jsx(
  "div",
  {
    style: {
      height: "40px",
      // This is to match the chevron's __next40pxDefaultSize
      width: "100%",
      display: "flex",
      justifyContent: "space-between",
      alignItems: "center"
    },
    children
  }
) });
function useCardHeader(layout) {
  const { isOpened, isCollapsible } = layout;
  const [isOpen, setIsOpen] = useState(isOpened);
  const [touched, setTouched] = useState(false);
  useEffect(() => {
    setIsOpen(isOpened);
  }, [isOpened]);
  const toggle = useCallback(() => {
    if (isOpen) {
      setTouched(true);
    }
    setIsOpen((prev) => !prev);
  }, [isOpen]);
  const CollapsibleCardHeader = useCallback(
    ({
      children,
      ...props
    }) => /* @__PURE__ */ jsxs(
      OriginalCardHeader,
      {
        ...props,
        onClick: toggle,
        style: {
          cursor: "pointer",
          ...props.style
        },
        isBorderless: true,
        children: [
          /* @__PURE__ */ jsx(
            "div",
            {
              style: {
                width: "100%",
                display: "flex",
                justifyContent: "space-between",
                alignItems: "center"
              },
              children
            }
          ),
          /* @__PURE__ */ jsx(
            Button,
            {
              __next40pxDefaultSize: true,
              variant: "tertiary",
              icon: isOpen ? chevronUp : chevronDown,
              "aria-expanded": isOpen,
              "aria-label": isOpen ? "Collapse" : "Expand"
            }
          )
        ]
      }
    ),
    [toggle, isOpen]
  );
  const effectiveIsOpen = isCollapsible ? isOpen : true;
  const CardHeaderComponent = isCollapsible ? CollapsibleCardHeader : NonCollapsibleCardHeader;
  return {
    isOpen: effectiveIsOpen,
    CardHeader: CardHeaderComponent,
    touched,
    setTouched
  };
}
function isSummaryFieldVisible(summaryField, summaryConfig, isOpen) {
  if (!summaryConfig || Array.isArray(summaryConfig) && summaryConfig.length === 0) {
    return false;
  }
  const summaryConfigArray = Array.isArray(summaryConfig) ? summaryConfig : [summaryConfig];
  const fieldConfig = summaryConfigArray.find((config) => {
    if (typeof config === "string") {
      return config === summaryField.id;
    }
    if (typeof config === "object" && "id" in config) {
      return config.id === summaryField.id;
    }
    return false;
  });
  if (!fieldConfig) {
    return false;
  }
  if (typeof fieldConfig === "string") {
    return true;
  }
  if (typeof fieldConfig === "object" && "visibility" in fieldConfig) {
    return fieldConfig.visibility === "always" || fieldConfig.visibility === "when-collapsed" && !isOpen;
  }
  return true;
}
function FormCardField({
  data,
  field,
  onChange,
  hideLabelFromVision,
  validity
}) {
  const { fields } = useContext(DataFormContext);
  const layout = field.layout;
  const cardBodyRef = useRef(null);
  const form = useMemo(
    () => ({
      layout: DEFAULT_LAYOUT,
      fields: field.children ?? []
    }),
    [field]
  );
  const { isOpen, CardHeader, touched, setTouched } = useCardHeader(layout);
  const handleBlur = useCallback(() => {
    setTouched(true);
  }, [setTouched]);
  useReportValidity(cardBodyRef, isOpen && touched);
  const summaryFields = getSummaryFields(layout.summary, fields);
  const visibleSummaryFields = summaryFields.filter(
    (summaryField) => isSummaryFieldVisible(summaryField, layout.summary, isOpen)
  );
  const invalidCount = countInvalidFields(validity);
  const showValidationBadge = touched && invalidCount > 0 && layout.isCollapsible;
  const validationBadge = showValidationBadge ? /* @__PURE__ */ jsx(Badge, { intent: "high", children: sprintf(
    /* translators: %d: Number of fields that need attention */
    _n(
      "%d field needs attention",
      "%d fields need attention",
      invalidCount
    ),
    invalidCount
  ) }) : null;
  const sizeCard = {
    blockStart: "medium",
    blockEnd: "medium",
    inlineStart: "medium",
    inlineEnd: "medium"
  };
  if (!!field.children) {
    const withHeader2 = !!field.label && layout.withHeader;
    const sizeCardBody2 = {
      blockStart: withHeader2 ? "none" : "medium",
      blockEnd: "medium",
      inlineStart: "medium",
      inlineEnd: "medium"
    };
    return /* @__PURE__ */ jsxs(Card, { className: "dataforms-layouts-card__field", size: sizeCard, children: [
      withHeader2 && /* @__PURE__ */ jsxs(CardHeader, { className: "dataforms-layouts-card__field-header", children: [
        /* @__PURE__ */ jsx("span", { className: "dataforms-layouts-card__field-header-label", children: field.label }),
        validationBadge,
        visibleSummaryFields.length > 0 && layout.withHeader && /* @__PURE__ */ jsx("div", { className: "dataforms-layouts-card__field-summary", children: visibleSummaryFields.map(
          (summaryField) => /* @__PURE__ */ jsx(
            summaryField.render,
            {
              item: data,
              field: summaryField
            },
            summaryField.id
          )
        ) })
      ] }),
      (isOpen || !withHeader2) && // If it doesn't have a header, keep it open.
      // Otherwise, the card will not be visible.
      /* @__PURE__ */ jsxs(
        CardBody,
        {
          size: sizeCardBody2,
          className: "dataforms-layouts-card__field-control",
          ref: cardBodyRef,
          onBlur: handleBlur,
          children: [
            field.description && /* @__PURE__ */ jsx("div", { className: "dataforms-layouts-card__field-description", children: field.description }),
            /* @__PURE__ */ jsx(
              DataFormLayout,
              {
                data,
                form,
                onChange,
                validity: validity?.children
              }
            )
          ]
        }
      )
    ] });
  }
  const fieldDefinition = fields.find(
    (fieldDef) => fieldDef.id === field.id
  );
  if (!fieldDefinition || !fieldDefinition.Edit) {
    return null;
  }
  const RegularLayout = getFormFieldLayout("regular")?.component;
  if (!RegularLayout) {
    return null;
  }
  const withHeader = !!fieldDefinition.label && layout.withHeader;
  const sizeCardBody = {
    blockStart: withHeader ? "none" : "medium",
    blockEnd: "medium",
    inlineStart: "medium",
    inlineEnd: "medium"
  };
  return /* @__PURE__ */ jsxs(Card, { className: "dataforms-layouts-card__field", size: sizeCard, children: [
    withHeader && /* @__PURE__ */ jsxs(CardHeader, { className: "dataforms-layouts-card__field-header", children: [
      /* @__PURE__ */ jsx("span", { className: "dataforms-layouts-card__field-header-label", children: fieldDefinition.label }),
      validationBadge,
      visibleSummaryFields.length > 0 && layout.withHeader && /* @__PURE__ */ jsx("div", { className: "dataforms-layouts-card__field-summary", children: visibleSummaryFields.map((summaryField) => /* @__PURE__ */ jsx(
        summaryField.render,
        {
          item: data,
          field: summaryField
        },
        summaryField.id
      )) })
    ] }),
    (isOpen || !withHeader) && // If it doesn't have a header, keep it open.
    // Otherwise, the card will not be visible.
    /* @__PURE__ */ jsx(
      CardBody,
      {
        size: sizeCardBody,
        className: "dataforms-layouts-card__field-control",
        ref: cardBodyRef,
        onBlur: handleBlur,
        children: /* @__PURE__ */ jsx(
          RegularLayout,
          {
            data,
            field,
            onChange,
            hideLabelFromVision: hideLabelFromVision || withHeader,
            validity
          }
        )
      }
    )
  ] });
}
export {
  FormCardField as default,
  useCardHeader
};
