// @wordpress/ui/build-module/badge/badge.mjs
// packages/ui/src/badge/badge.tsx
import { forwardRef } from "@wordpress/element";
import { Box } from "../box/index.mjs";
import { jsx } from "react/jsx-runtime";
var DEFAULT_RENDER = (props) => /* @__PURE__ */ jsx("span", { ...props });
var getIntentStyles = (intent) => {
  switch (intent) {
    case "high":
      return {
        backgroundColor: "error",
        color: "error"
      };
    case "medium":
      return {
        backgroundColor: "warning",
        color: "warning"
      };
    case "low":
      return {
        backgroundColor: "caution",
        color: "caution"
      };
    case "stable":
      return {
        backgroundColor: "success",
        color: "success"
      };
    case "informational":
      return {
        backgroundColor: "info",
        color: "info"
      };
    case "draft":
      return {
        backgroundColor: "neutral-weak",
        color: "neutral"
      };
    case "none":
    default:
      return {
        backgroundColor: "neutral",
        color: "neutral-weak"
      };
  }
};
var Badge = forwardRef(function Badge2({ children, intent = "none", render = DEFAULT_RENDER, ...props }, ref) {
  const intentStyles = getIntentStyles(intent);
  return /* @__PURE__ */ jsx(
    Box,
    {
      ...intentStyles,
      padding: { inline: "xs", block: "2xs" },
      borderRadius: "lg",
      render,
      style: {
        fontFamily: "var(--wpds-font-family-body)",
        fontSize: "var(--wpds-font-size-sm)",
        fontWeight: "var(--wpds-font-weight-regular)",
        lineHeight: "var(--wpds-font-line-height-xs)",
        ...props.style
      },
      ref,
      children
    }
  );
});
export {
  Badge
};
