// @wordpress/dataviews/build/components/dataform-layouts/panel/index.cjs
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// packages/dataviews/src/components/dataform-layouts/panel/index.tsx
var panel_exports = {};
__export(panel_exports, {
  default: () => FormPanelField
});
module.exports = __toCommonJS(panel_exports);
var import_clsx = __toESM(require("clsx"));
var import_components = require("@wordpress/components");
var import_element = require("@wordpress/element");
var import_icons = require("@wordpress/icons");
var import_ui = require("@wordpress/ui");
var import_dataform_context = __toESM(require("../../dataform-context/index.cjs"));
var import_dropdown = __toESM(require("./dropdown.cjs"));
var import_modal = __toESM(require("./modal.cjs"));
var import_get_summary_fields = require("../get-summary-fields.cjs");
var import_jsx_runtime = require("react/jsx-runtime");
function getFirstValidationError(validity) {
  if (!validity) {
    return void 0;
  }
  const validityRules = Object.keys(validity).filter(
    (key) => key !== "children"
  );
  for (const key of validityRules) {
    const rule = validity[key];
    if (rule === void 0) {
      continue;
    }
    if (rule.type === "invalid") {
      if (rule.message) {
        return rule.message;
      }
      if (key === "required") {
        return "A required field is empty";
      }
      return "Unidentified validation error";
    }
  }
  if (validity.children) {
    for (const childValidity of Object.values(validity.children)) {
      const childError = getFirstValidationError(childValidity);
      if (childError) {
        return childError;
      }
    }
  }
  return void 0;
}
var getFieldDefinition = (field, fields) => {
  const fieldDefinition = fields.find((_field) => _field.id === field.id);
  if (!fieldDefinition) {
    return fields.find((_field) => {
      if (!!field.children) {
        const simpleChildren = field.children.filter(
          (child) => !child.children
        );
        if (simpleChildren.length === 0) {
          return false;
        }
        return _field.id === simpleChildren[0].id;
      }
      return _field.id === field.id;
    });
  }
  return fieldDefinition;
};
var getFieldDefinitionAndSummaryFields = (layout, field, fields) => {
  const summaryFields = (0, import_get_summary_fields.getSummaryFields)(layout.summary, fields);
  const fieldDefinition = getFieldDefinition(field, fields);
  if (summaryFields.length === 0) {
    return {
      summaryFields: fieldDefinition ? [fieldDefinition] : [],
      fieldDefinition
    };
  }
  return {
    summaryFields,
    fieldDefinition
  };
};
function FormPanelField({
  data,
  field,
  onChange,
  validity
}) {
  const { fields } = (0, import_element.useContext)(import_dataform_context.default);
  const layout = field.layout;
  const [popoverAnchor, setPopoverAnchor] = (0, import_element.useState)(
    null
  );
  const [touched, setTouched] = (0, import_element.useState)(false);
  const handleClose = () => setTouched(true);
  const { fieldDefinition, summaryFields } = getFieldDefinitionAndSummaryFields(layout, field, fields);
  if (!fieldDefinition) {
    return null;
  }
  const labelPosition = layout.labelPosition;
  const errorMessage = getFirstValidationError(validity);
  const showError = touched && !!errorMessage;
  const labelClassName = (0, import_clsx.default)(
    "dataforms-layouts-panel__field-label",
    `dataforms-layouts-panel__field-label--label-position-${labelPosition}`,
    { "has-error": showError }
  );
  const fieldLabel = !!field.children ? field.label : fieldDefinition?.label;
  const labelContent = showError ? /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_components.Tooltip, { text: errorMessage, placement: "top", children: /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
    import_ui.Stack,
    {
      direction: "row",
      gap: "xs",
      className: "dataforms-layouts-panel__field-label-error-content",
      justify: "flex-start",
      children: [
        /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_components.Icon, { icon: import_icons.error, size: 16 }),
        /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_jsx_runtime.Fragment, { children: fieldLabel })
      ]
    }
  ) }) : fieldLabel;
  const renderedControl = layout.openAs === "modal" ? /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    import_modal.default,
    {
      data,
      field,
      onChange,
      labelPosition,
      summaryFields,
      fieldDefinition,
      onClose: handleClose,
      touched
    }
  ) : /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    import_dropdown.default,
    {
      data,
      field,
      onChange,
      validity,
      labelPosition,
      summaryFields,
      fieldDefinition,
      popoverAnchor,
      onClose: handleClose,
      touched
    }
  );
  if (labelPosition === "top") {
    return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
      import_ui.Stack,
      {
        direction: "column",
        className: "dataforms-layouts-panel__field",
        children: [
          /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
            "div",
            {
              className: labelClassName,
              style: { paddingBottom: 0 },
              children: labelContent
            }
          ),
          /* @__PURE__ */ (0, import_jsx_runtime.jsx)("div", { className: "dataforms-layouts-panel__field-control", children: renderedControl })
        ]
      }
    );
  }
  if (labelPosition === "none") {
    return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
      import_ui.Stack,
      {
        direction: "row",
        gap: "xs",
        className: "dataforms-layouts-panel__field dataforms-layouts-panel__field--label-position-none",
        children: [
          showError && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_components.Tooltip, { text: errorMessage, placement: "top", children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
            import_components.Icon,
            {
              className: "dataforms-layouts-panel__field-label-error-content",
              icon: import_icons.error,
              size: 16
            }
          ) }),
          /* @__PURE__ */ (0, import_jsx_runtime.jsx)("div", { className: "dataforms-layouts-panel__field-control", children: renderedControl })
        ]
      }
    );
  }
  return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
    import_ui.Stack,
    {
      direction: "row",
      gap: "xs",
      ref: setPopoverAnchor,
      className: "dataforms-layouts-panel__field",
      children: [
        /* @__PURE__ */ (0, import_jsx_runtime.jsx)("div", { className: labelClassName, children: labelContent }),
        /* @__PURE__ */ (0, import_jsx_runtime.jsx)("div", { className: "dataforms-layouts-panel__field-control", children: renderedControl })
      ]
    }
  );
}
