// @wordpress/dataviews/build/components/dataform-layouts/panel/summary-button.cjs
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// packages/dataviews/src/components/dataform-layouts/panel/summary-button.tsx
var summary_button_exports = {};
__export(summary_button_exports, {
  default: () => summary_button_default
});
module.exports = __toCommonJS(summary_button_exports);
var import_components = require("@wordpress/components");
var import_i18n = require("@wordpress/i18n");
var import_jsx_runtime = require("react/jsx-runtime");
function SummaryButton({
  summaryFields,
  data,
  labelPosition,
  fieldLabel,
  disabled,
  onClick,
  "aria-expanded": ariaExpanded
}) {
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    import_components.Button,
    {
      className: "dataforms-layouts-panel__summary-button",
      size: "compact",
      variant: ["none", "top"].includes(labelPosition) ? "link" : "tertiary",
      "aria-expanded": ariaExpanded,
      "aria-label": (0, import_i18n.sprintf)(
        // translators: %s: Field name.
        (0, import_i18n._x)("Edit %s", "field"),
        fieldLabel || ""
      ),
      onClick,
      disabled,
      accessibleWhenDisabled: true,
      style: summaryFields.length > 1 ? {
        minHeight: "auto",
        height: "auto",
        alignItems: "flex-start"
      } : void 0,
      children: summaryFields.length > 1 ? /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
        "div",
        {
          style: {
            display: "flex",
            flexDirection: "column",
            alignItems: "flex-start",
            width: "100%",
            gap: "2px"
          },
          children: summaryFields.map((summaryField) => /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
            "div",
            {
              style: { width: "100%" },
              children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
                summaryField.render,
                {
                  item: data,
                  field: summaryField
                }
              )
            },
            summaryField.id
          ))
        }
      ) : summaryFields.map((summaryField) => /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
        summaryField.render,
        {
          item: data,
          field: summaryField
        },
        summaryField.id
      ))
    }
  );
}
var summary_button_default = SummaryButton;
