// @wordpress/dataviews/build-module/components/dataviews-footer/index.mjs
// packages/dataviews/src/components/dataviews-footer/index.tsx
import { useContext } from "@wordpress/element";
import { Stack } from "@wordpress/ui";
import DataViewsContext from "../dataviews-context/index.mjs";
import DataViewsPagination from "../dataviews-pagination/index.mjs";
import {
  BulkActionsFooter,
  useSomeItemHasAPossibleBulkAction
} from "../dataviews-bulk-actions/index.mjs";
import { LAYOUT_GRID, LAYOUT_TABLE } from "../../constants.mjs";
import { jsx, jsxs } from "react/jsx-runtime";
var EMPTY_ARRAY = [];
function DataViewsFooter() {
  const {
    view,
    paginationInfo: { totalItems = 0, totalPages },
    data,
    actions = EMPTY_ARRAY
  } = useContext(DataViewsContext);
  const hasBulkActions = useSomeItemHasAPossibleBulkAction(actions, data) && [LAYOUT_TABLE, LAYOUT_GRID].includes(view.type);
  if (!totalItems || !totalPages || totalPages <= 1 && !hasBulkActions) {
    return null;
  }
  return !!totalItems && /* @__PURE__ */ jsxs(
    Stack,
    {
      direction: "row",
      justify: "end",
      align: "center",
      className: "dataviews-footer",
      gap: "xs",
      children: [
        hasBulkActions && /* @__PURE__ */ jsx(BulkActionsFooter, {}),
        /* @__PURE__ */ jsx(DataViewsPagination, {})
      ]
    }
  );
}
export {
  DataViewsFooter as default
};
