// @wordpress/ui/build/badge/badge.cjs
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// packages/ui/src/badge/badge.tsx
var badge_exports = {};
__export(badge_exports, {
  Badge: () => Badge
});
module.exports = __toCommonJS(badge_exports);
var import_element = require("@wordpress/element");
var import_box = require("../box/index.cjs");
var import_jsx_runtime = require("react/jsx-runtime");
var DEFAULT_RENDER = (props) => /* @__PURE__ */ (0, import_jsx_runtime.jsx)("span", { ...props });
var getIntentStyles = (intent) => {
  switch (intent) {
    case "high":
      return {
        backgroundColor: "error",
        color: "error"
      };
    case "medium":
      return {
        backgroundColor: "warning",
        color: "warning"
      };
    case "low":
      return {
        backgroundColor: "caution",
        color: "caution"
      };
    case "stable":
      return {
        backgroundColor: "success",
        color: "success"
      };
    case "informational":
      return {
        backgroundColor: "info",
        color: "info"
      };
    case "draft":
      return {
        backgroundColor: "neutral-weak",
        color: "neutral"
      };
    case "none":
    default:
      return {
        backgroundColor: "neutral",
        color: "neutral-weak"
      };
  }
};
var Badge = (0, import_element.forwardRef)(function Badge2({ children, intent = "none", render = DEFAULT_RENDER, ...props }, ref) {
  const intentStyles = getIntentStyles(intent);
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    import_box.Box,
    {
      ...intentStyles,
      padding: { inline: "xs", block: "2xs" },
      borderRadius: "lg",
      render,
      style: {
        fontFamily: "var(--wpds-font-family-body)",
        fontSize: "var(--wpds-font-size-sm)",
        fontWeight: "var(--wpds-font-weight-regular)",
        lineHeight: "var(--wpds-font-line-height-xs)",
        ...props.style
      },
      ref,
      children
    }
  );
});
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  Badge
});
