// @wordpress/dataviews/build-module/components/dataform-layouts/panel/index.mjs
// packages/dataviews/src/components/dataform-layouts/panel/index.tsx
import clsx from "clsx";
import { Icon, Tooltip } from "@wordpress/components";
import { useState, useContext } from "@wordpress/element";
import { error as errorIcon } from "@wordpress/icons";
import { Stack } from "@wordpress/ui";
import DataFormContext from "../../dataform-context/index.mjs";
import PanelDropdown from "./dropdown.mjs";
import PanelModal from "./modal.mjs";
import { getSummaryFields } from "../get-summary-fields.mjs";
import { Fragment, jsx, jsxs } from "react/jsx-runtime";
function getFirstValidationError(validity) {
  if (!validity) {
    return void 0;
  }
  const validityRules = Object.keys(validity).filter(
    (key) => key !== "children"
  );
  for (const key of validityRules) {
    const rule = validity[key];
    if (rule === void 0) {
      continue;
    }
    if (rule.type === "invalid") {
      if (rule.message) {
        return rule.message;
      }
      if (key === "required") {
        return "A required field is empty";
      }
      return "Unidentified validation error";
    }
  }
  if (validity.children) {
    for (const childValidity of Object.values(validity.children)) {
      const childError = getFirstValidationError(childValidity);
      if (childError) {
        return childError;
      }
    }
  }
  return void 0;
}
var getFieldDefinition = (field, fields) => {
  const fieldDefinition = fields.find((_field) => _field.id === field.id);
  if (!fieldDefinition) {
    return fields.find((_field) => {
      if (!!field.children) {
        const simpleChildren = field.children.filter(
          (child) => !child.children
        );
        if (simpleChildren.length === 0) {
          return false;
        }
        return _field.id === simpleChildren[0].id;
      }
      return _field.id === field.id;
    });
  }
  return fieldDefinition;
};
var getFieldDefinitionAndSummaryFields = (layout, field, fields) => {
  const summaryFields = getSummaryFields(layout.summary, fields);
  const fieldDefinition = getFieldDefinition(field, fields);
  if (summaryFields.length === 0) {
    return {
      summaryFields: fieldDefinition ? [fieldDefinition] : [],
      fieldDefinition
    };
  }
  return {
    summaryFields,
    fieldDefinition
  };
};
function FormPanelField({
  data,
  field,
  onChange,
  validity
}) {
  const { fields } = useContext(DataFormContext);
  const layout = field.layout;
  const [popoverAnchor, setPopoverAnchor] = useState(
    null
  );
  const [touched, setTouched] = useState(false);
  const handleClose = () => setTouched(true);
  const { fieldDefinition, summaryFields } = getFieldDefinitionAndSummaryFields(layout, field, fields);
  if (!fieldDefinition) {
    return null;
  }
  const labelPosition = layout.labelPosition;
  const errorMessage = getFirstValidationError(validity);
  const showError = touched && !!errorMessage;
  const labelClassName = clsx(
    "dataforms-layouts-panel__field-label",
    `dataforms-layouts-panel__field-label--label-position-${labelPosition}`,
    { "has-error": showError }
  );
  const fieldLabel = !!field.children ? field.label : fieldDefinition?.label;
  const labelContent = showError ? /* @__PURE__ */ jsx(Tooltip, { text: errorMessage, placement: "top", children: /* @__PURE__ */ jsxs(
    Stack,
    {
      direction: "row",
      gap: "xs",
      className: "dataforms-layouts-panel__field-label-error-content",
      justify: "flex-start",
      children: [
        /* @__PURE__ */ jsx(Icon, { icon: errorIcon, size: 16 }),
        /* @__PURE__ */ jsx(Fragment, { children: fieldLabel })
      ]
    }
  ) }) : fieldLabel;
  const renderedControl = layout.openAs === "modal" ? /* @__PURE__ */ jsx(
    PanelModal,
    {
      data,
      field,
      onChange,
      labelPosition,
      summaryFields,
      fieldDefinition,
      onClose: handleClose,
      touched
    }
  ) : /* @__PURE__ */ jsx(
    PanelDropdown,
    {
      data,
      field,
      onChange,
      validity,
      labelPosition,
      summaryFields,
      fieldDefinition,
      popoverAnchor,
      onClose: handleClose,
      touched
    }
  );
  if (labelPosition === "top") {
    return /* @__PURE__ */ jsxs(
      Stack,
      {
        direction: "column",
        className: "dataforms-layouts-panel__field",
        children: [
          /* @__PURE__ */ jsx(
            "div",
            {
              className: labelClassName,
              style: { paddingBottom: 0 },
              children: labelContent
            }
          ),
          /* @__PURE__ */ jsx("div", { className: "dataforms-layouts-panel__field-control", children: renderedControl })
        ]
      }
    );
  }
  if (labelPosition === "none") {
    return /* @__PURE__ */ jsxs(
      Stack,
      {
        direction: "row",
        gap: "xs",
        className: "dataforms-layouts-panel__field dataforms-layouts-panel__field--label-position-none",
        children: [
          showError && /* @__PURE__ */ jsx(Tooltip, { text: errorMessage, placement: "top", children: /* @__PURE__ */ jsx(
            Icon,
            {
              className: "dataforms-layouts-panel__field-label-error-content",
              icon: errorIcon,
              size: 16
            }
          ) }),
          /* @__PURE__ */ jsx("div", { className: "dataforms-layouts-panel__field-control", children: renderedControl })
        ]
      }
    );
  }
  return /* @__PURE__ */ jsxs(
    Stack,
    {
      direction: "row",
      gap: "xs",
      ref: setPopoverAnchor,
      className: "dataforms-layouts-panel__field",
      children: [
        /* @__PURE__ */ jsx("div", { className: labelClassName, children: labelContent }),
        /* @__PURE__ */ jsx("div", { className: "dataforms-layouts-panel__field-control", children: renderedControl })
      ]
    }
  );
}
export {
  FormPanelField as default
};
